operator: Implement handle_deleted()

vatsa287 opened the pull request:
This PR implements deletion of storage volume which will be called
when kubectl delete -f is performed on the storage.yaml file. It checks
if pvs provisioned from that volume is zero or not and then proceeds to
delete the volume.
    
Changes:
- New class variables in storage_list.StorageUnit to make its use when
  obj is deleted.
- Includes storage_list and utils from cli into kadalu image FS.
- New ClusterRole 'pod-exec' to allow kadalu-operator to make exec into pods
  which has been binded into 'kadalu-operator'
- Delete ClusterRole 'pod-exec' in cleanup file.
- Renamed methods 'execute' which is defined in both kadalulib and utils.

Fixes: #326 

Signed-off-by: Shree Vatsa N <vatsa@kadalu.io>

aravindavk commented on operator/main.py:
```diff
@@ -464,22 +466,203 @@ def handle_modified(core_v1_client, obj):
+
     logging.warning(logf(
-        "DELETED handle called, but not implemented"
+        "Delete {} requested".format(volname)
```
Follow structured logging format. `logf(STATIC_MESSAGE, key1=value1, key2=value2...)`

Above log will become,

```
logging.warning(logf(
    "Delete requested",
    volname=volname
))
```

aravindavk commented on operator/main.py:
```diff
@@ -464,22 +466,203 @@ def handle_modified(core_v1_client, obj):
+            template(filename, namespace=NAMESPACE, volname=volname)
+            lib_execute(KUBECTL_CMD, DELETE_CMD, "-f", filename)
+            logging.info(logf("Deleted Service", volname=volname, manifest=filename))
+
```
Break after this.

It should be fine to iterate all in this case, but with the `break` you can come out of loop early.

aravindavk commented on operator/main.py:
```diff
@@ -464,22 +466,203 @@ def handle_modified(core_v1_client, obj):
+    for storage in storages:
+        if storage.storage_name == volname and storage.pv_count != 0:
+            logging.warning(logf(
+                "Number of PVs from {} is not zero".format(volname)
```
Same comment as above for the format. Also mention about what is failed. For example

```
logf(
    "Storage delete failed. Storage is not empty",
    number_of_pvs=storage.pv_count,
    storage=volname
)
```

aravindavk commented on operator/main.py:
```diff
@@ -464,22 +466,203 @@ def handle_modified(core_v1_client, obj):
+
+    return storages
+
+def set_num_pvs(storages, volname):
```
Change this to `get_num_pvs` and accept `storage`

aravindavk commented on operator/main.py:
```diff
@@ -464,22 +466,203 @@ def handle_modified(core_v1_client, obj):
+            pv_stats_parts = parts[1].strip().split("|")
+            storage.pv_count = int(pv_stats_parts[0])
+        except utils.CommandError as err:
+            print("Failed to get size details of the "
```
Use log instead of print here.

aravindavk commented on operator/main.py:
```diff
@@ -464,22 +466,203 @@ def handle_modified(core_v1_client, obj):
+            print()
+        except FileNotFoundError:
+            # remove 'args' : args not defined
+            utils.kubectl_cmd_help(args.kubectl_cmd)
```
Change this to `log`. This exception will not happen inside Operator container since `kubectl` binary will be available always.

aravindavk commented on operator/main.py:
```diff
@@ -464,22 +466,203 @@ def handle_modified(core_v1_client, obj):
+    except FileNotFoundError:
+        utils.kubectl_cmd_help(args.kubectl_cmd)
+
+    set_num_pvs(storages, volname)
```
Rename as mentioned above. and move this inside for loop below.

aravindavk commented on operator/main.py:
```diff
@@ -464,22 +466,203 @@ def handle_modified(core_v1_client, obj):
+
+    set_num_pvs(storages, volname)
+
+    for storage in storages:
```
Use below logic to keep the following if statements simple

```
if storage.storage_name != volname:
    continue
```

aravindavk commented on operator/main.py:
```diff
@@ -464,22 +466,203 @@ def handle_modified(core_v1_client, obj):
+        KADALU_CONFIG_MAP, NAMESPACE)
+
+    # For each brick, add brick path and node id
+    bricks = obj["spec"]["storage"]
```
Except `name` nothing else will be available in `obj`. You can skip this bricks part

aravindavk commented on operator/main.py:
```diff
@@ -464,22 +466,203 @@ def handle_modified(core_v1_client, obj):
+    voltype = obj["spec"]["type"]
+    volume_id = obj["spec"]["volume_id"]
+
+    data = {
```
Not required

aravindavk commented on operator/main.py:
```diff
@@ -464,22 +466,203 @@ def handle_modified(core_v1_client, obj):
+    Volinfo of new hosting Volume is generated and updated to ConfigMap
+    """
+    volname = obj["metadata"]["name"]
+    voltype = obj["spec"]["type"]
```
voltype and volume_id not available in obj

aravindavk commented on operator/main.py:
```diff
@@ -464,22 +466,203 @@ def handle_modified(core_v1_client, obj):
+        })
+
+    volinfo_file = "%s.info" % volname
+    configmap_data.data[volinfo_file] = ""
```
`del configmap_data.data[volinfo_file]`

aravindavk commented on operator/main.py:
```diff
@@ -464,22 +466,203 @@ def handle_modified(core_v1_client, obj):
+                      volname=volname))
+
+
+def delete_server_pods(obj):
```
Accept `storage` instead of `obj` or along with `obj` so that it will be easier to get server pod/statefulset details.

aravindavk commented on operator/main.py:
```diff
@@ -464,22 +466,203 @@ def handle_modified(core_v1_client, obj):
+    """
+    # Delete server pod
+    volname = obj["metadata"]["name"]
+    voltype = obj["spec"]["type"]
```
`voltype` not available in `obj` get it from `storage`

aravindavk commented on operator/main.py:
```diff
@@ -464,22 +466,203 @@ def handle_modified(core_v1_client, obj):
+
+    try:
+        resp = utils.execute(cmd)
+        storages = storage_list.list_storages(resp.stdout, args)
```
`get_storages` function defined above can be used here.

aravindavk commented on operator/main.py:
```diff
@@ -464,22 +466,203 @@ def handle_modified(core_v1_client, obj):
+        #list_storages does not make use of use 'args'
+        storages = storage_list.list_storages(resp.stdout, None)
+    except utils.CommandError as err:
+        utils.command_error(cmd1, err.stderr)
```
Do not print error inside Operator, will not be visible since it is running as daemon, use `log`

vatsa287 commented:
@aravindavk Thanks a lot , will make changes

vatsa287 commented on operator/main.py:
```diff
@@ -464,22 +466,203 @@ def handle_modified(core_v1_client, obj):
+def get_storages():
+    """Parse list of Storages from JSON input"""
+
+    storage = storage_list.Storage()
```
Will 'storage' be a class object with only default values set or not? 'storage' is already being instantiated inside list_storages() but its not returning 'storage' or 'storage_unit' but only 'storages'

vatsa287 committed:
operator: Implement handle_deleted()

This PR implements deletion of storage volume which will be called
when kubectl delete -f is performed on the storage.yaml file. It checks
if pvs provisioned from that volume is zero or not and then proceeds to
delete the volume.

Changes:
- New class variables in storage_list.StorageUnit to make its use when
  obj is deleted.
- Includes storage_list and utils from cli into kadalu image FS.
- New ClusterRole 'pod-exec' to allow kadalu-operator to make exec into pods
  which has been binded into 'kadalu-operator'
- Delete ClusterRole 'pod-exec' in cleanup file.
- Renamed methods 'execute' which is defined in both kadalulib and utils.

Signed-off-by: Shree Vatsa N <vatsa@kadalu.io>

amarts commented on operator/main.py:
```diff
@@ -464,22 +466,178 @@ def handle_modified(core_v1_client, obj):
     ))
 
+    # Intialise Variables
+    storage = Storage()
```
this gets overwritten.. is that OK?

vatsa287 replied:
I was not able test it properly since the PV was not binding. Ill change the default values and check again for storage.pv_count

vatsa287 commented on operator/Dockerfile:
```diff
@@ -21,7 +21,9 @@ COPY templates/storageclass-kadalu.replica1.yaml.j2  /kadalu/templates/storagecl
 COPY templates/external-storageclass.yaml.j2  /kadalu/templates/external-storageclass.yaml.j2
-COPY lib/kadalulib.py                /kadalu/kadalulib.py
+COPY lib/kadalulib.py                    /kadalu/kadalulib.py
+COPY cli/kubectl_kadalu/storage_list.py  /kadalu/storage_list.py
```
Getting import error even though I have storage_list and utils in operator's image? [resolved]

vatsa287 commented:
### TEST RUN
```bash
# When PV count is not 0
$ sudo kubectl logs -nkadalu operator-59f656cf56-2z8f4
[2020-12-04 18:44:36,574] WARNING [main - 599:handle_deleted] - Delete requested     volname=storage-pool-1
[2020-12-04 18:44:37,260] WARNING [main - 615:handle_deleted] - Storage delete failed. Storage is not empty  number_of_pvs=1 storage=storage-pool-1

$ sudo kubectl get pods -nkadalu
NAME                          READY   STATUS    RESTARTS   AGE
kadalu-csi-nodeplugin-nxwj4   3/3     Running   0          8m15s
kadalu-csi-provisioner-0      4/4     Running   0          8m14s
operator-59f656cf56-2z8f4     1/1     Running   0          8m22s
server-storage-pool-1-0-0     2/2     Running   0          4m25s

# When PV count is 0
$ sudo kubectl logs -nkadalu operator-59f656cf56-2z8f4
[2020-12-04 18:52:17,754] WARNING [main - 599:handle_deleted] - Delete requested     volname=storage-pool-1
/usr/local/lib/python3.9/site-packages/urllib3/connectionpool.py:1013: InsecureRequestWarning: Unverified HTTPS request is being made to host '10.96.0.1'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  warnings.warn(
[2020-12-04 18:52:18,904] DEBUG [rest - 228:request] - response body: {"kind":"ConfigMap","apiVersion":"v1","metadata":{"name":"kadalu-info","namespace":"kadalu","selfLink":"/api/v1/namespaces/kadalu/configmaps/kadalu-info","uid":"f7f55648-5e9b-4e53-ac59-2753df900dec","resourceVersion":"1356467","creationTimestamp":"2020-12-04T18:36:47Z","annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"v1\",\"data\":{\"uid\":\"2709a16b-88b5-49ab-b373-59891ccd8450\",\"volumes\":\"\"},\"kind\":\"ConfigMap\",\"metadata\":{\"annotations\":{},\"name\":\"kadalu-info\",\"namespace\":\"kadalu\"}}\n"}},"data":{"storage-pool-1.info":"{\"namespace\": \"kadalu\", \"kadalu_version\": \"devel\", \"volname\": \"storage-pool-1\", \"volume_id\": \"33a2347c-3660-11eb-a2b3-0242ac110004\", \"type\": \"Replica1\", \"bricks\": [{\"brick_path\": \"/bricks/storage-pool-1/data/brick\", \"kube_hostname\": \"minikube\", \"node\": \"server-storage-pool-1-0-0.storage-pool-1\", \"node_id\": \"33a23724-3660-11eb-a2b3-0242ac110004\", \"host_brick_path\": \"\", \"brick_device\": \"/home/vatsa/disk3\", \"pvc_name\": \"\", \"brick_device_dir\": \"/home/vatsa\", \"brick_index\": 0}], \"options\": {}}","uid":"2709a16b-88b5-49ab-b373-59891ccd8450","volumes":""}}

/usr/local/lib/python3.9/site-packages/urllib3/connectionpool.py:1013: InsecureRequestWarning: Unverified HTTPS request is being made to host '10.96.0.1'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  warnings.warn(
[2020-12-04 18:52:18,919] DEBUG [rest - 228:request] - response body: {"kind":"ConfigMap","apiVersion":"v1","metadata":{"name":"kadalu-info","namespace":"kadalu","selfLink":"/api/v1/namespaces/kadalu/configmaps/kadalu-info","uid":"f7f55648-5e9b-4e53-ac59-2753df900dec","resourceVersion":"1356467","creationTimestamp":"2020-12-04T18:36:47Z","annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"v1\",\"data\":{\"uid\":\"2709a16b-88b5-49ab-b373-59891ccd8450\",\"volumes\":\"\"},\"kind\":\"ConfigMap\",\"metadata\":{\"annotations\":{},\"name\":\"kadalu-info\",\"namespace\":\"kadalu\"}}\n"}},"data":{"storage-pool-1.info":"{\"namespace\": \"kadalu\", \"kadalu_version\": \"devel\", \"volname\": \"storage-pool-1\", \"volume_id\": \"33a2347c-3660-11eb-a2b3-0242ac110004\", \"type\": \"Replica1\", \"bricks\": [{\"brick_path\": \"/bricks/storage-pool-1/data/brick\", \"kube_hostname\": \"minikube\", \"node\": \"server-storage-pool-1-0-0.storage-pool-1\", \"node_id\": \"33a23724-3660-11eb-a2b3-0242ac110004\", \"host_brick_path\": \"\", \"brick_device\": \"/home/vatsa/disk3\", \"pvc_name\": \"\", \"brick_device_dir\": \"/home/vatsa\", \"brick_index\": 0}], \"options\": {}}","uid":"2709a16b-88b5-49ab-b373-59891ccd8450","volumes":""}}

[2020-12-04 18:52:18,920] INFO [main - 487:delete_config_map] - Deleted configmap    name=kadalu-info volname=storage-pool-1
[2020-12-04 18:52:19,212] INFO [main - 537:delete_server_pods] - Deleted Server pod  volname=storage-pool-1 manifest=/kadalu/templates/server.yaml node=server-storage-pool-1-0-0.storage-pool-1
[2020-12-04 18:52:19,377] INFO [main - 632:handle_deleted] - Deleted Service     volname=storage-pool-1 manifest=/kadalu/templates/services.yaml

$ sudo kubectl get pods -nkadalu
NAME                          READY   STATUS    RESTARTS   AGE
kadalu-csi-nodeplugin-nxwj4   3/3     Running   0          16m
kadalu-csi-provisioner-0      4/4     Running   0          16m
operator-59f656cf56-2z8f4     1/1     Running   0          16m
```

amarts commented:
Looks good now, just one more change for pylint test.

```
diff --git a/cli/kubectl_kadalu/storage_list.py b/cli/kubectl_kadalu/storage_list.py
index 8b46de8..c290b3b 100644
--- a/cli/kubectl_kadalu/storage_list.py
+++ b/cli/kubectl_kadalu/storage_list.py
@@ -11,6 +11,7 @@ import utils
 # noqa # pylint: disable=useless-object-inheritance
 # noqa # pylint: disable=too-few-public-methods
 # noqa # pylint: disable=bad-option-value
+# noqa # pylint: disable=too-many-instance-attributes
 class StorageUnit(object):
     """Structure for Brick/Storage unit"""
     def __init__(self):
```
